staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/custom.go

```diff
@@ -0,0 +1,342 @@
+// cbor.Marshaler and likewise for the respective Unmarshaler interfaces. This will be removed in
+// favor of automatic transcoding between CBOR and JSON/text for these types.
+func RejectCustomMarshalers(v interface{}) error {
+	if v == nil {
+		return nil
+	}
+	rv := reflect.ValueOf(v)
+	if err := marshalerCache.getChecker(rv.Type()).check(rv, 128); err != nil {
+		return fmt.Errorf("unable to serialize %T: %w", v, err)
+	}
+	if err := unmarshalerCache.getChecker(rv.Type()).check(rv, 128); err != nil {
+		return fmt.Errorf("unable to serialize %T: %w", v, err)
+	}
+	return nil
+}
+
+// Recursion depth is limited as a basic mitigation against cyclic objects. Objects created by the
+// decoder shouldn't be able to contain cycles, but practically any object can be passed to the
+// encoder.
+var errMaxDepthExceeded = errors.New("object depth exceeds reasonable limits (possible cycle?)")
+
+var marshalerCache = checkers{
+	safeInterface: reflect.TypeFor[cbor.Marshaler](),
+	unsafeInterfaces: []reflect.Type{
+		reflect.TypeFor[json.Marshaler](),
+		reflect.TypeFor[encoding.TextMarshaler](),
+	},
+}
+
+var unmarshalerCache = checkers{
+	safeInterface: reflect.TypeFor[cbor.Unmarshaler](),
+	unsafeInterfaces: []reflect.Type{
+		reflect.TypeFor[json.Unmarshaler](),
+		reflect.TypeFor[encoding.TextUnmarshaler](),
+	},
+	assumeAddressableValues: true,
+}
+
+type checker struct {
+	check func(rv reflect.Value, depth int) error
+}
+
+// Having a single addressable checker for comparisons lets us prune and collapse parts of the
+// object traversal that are statically known to be safe.
+var noop = checker{
+	check: func(rv reflect.Value, depth int) error {
+		return nil
+	},
+}
+
+type checkers struct {
+	lock             sync.RWMutex
+	m                map[reflect.Type]**checker
+	safeInterface    reflect.Type
+	unsafeInterfaces []reflect.Type
+
+	assumeAddressableValues bool
+}
+
+func (cache *checkers) getChecker(rt reflect.Type) checker {
+	cache.lock.RLock()
+	c, ok := cache.m[rt]
+	cache.lock.RUnlock()
+	if ok {
+		return **c
+	}
+
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	return *cache.getCheckerLocked(rt)
+}
+
+func (cache *checkers) getCheckerLocked(rt reflect.Type) *checker {
+	if c, ok := cache.m[rt]; ok {
+		// This type was cached while waiting to acquire the lock.
+		return *c
+	}
+
+	// Store the cache entry now, before populating it, so that circular references between
+	// types don't overflow the call stack.
+	c := new(checker)
+	if cache.m == nil {
+		cache.m = make(map[reflect.Type]**checker)
+	}
+	cache.m[rt] = &c
+
+	// Take a nonreflective path for the unstructured container types. They're common and
+	// usually nested inside one another.
+	switch rt {
+	case reflect.TypeFor[map[string]interface{}](), reflect.TypeFor[[]interface{}]():
+		var nonreflect func(v interface{}, depth int) error
+		nonreflect = func(v interface{}, depth int) error {
```

Comment:
style nit (DRY)

```suggestion
		nonreflect := func(v interface{}, depth int) error {
```

Reply:
It needs the separate declaration to be able to recurse from inside the closure. I'll just pull it out to a top-level function, that seems like it'd be clearer anyway.